app/core/models.py

```diff
@@ -366,17 +373,26 @@ def check_cancel_requested(self):
         status = SubmissionRun.objects.values_list("status", flat=True).get(pk=self.id)
         return status == Status.CANCEL_PENDING
 
-    def mark_for_cancel(self):
+    def mark_for_cancel(self, remote=False):
         with transaction.atomic():
-            SubmissionRun.objects.filter(pk=self.id).update(
-                status=Status.CANCEL_PENDING
-            )
-            Evaluation.objects.filter(
-                submission_run_id=self.id, status=Status.PENDING
-            ).update(status=Status.CANCELLED)
-            Evaluation.objects.filter(
-                submission_run_id=self.id, status=Status.RUNNING
-            ).update(status=Status.CANCEL_PENDING)
+            SubmissionRun.objects.select_for_update().filter(pk=self.id)
+            submission_run = SubmissionRun.objects.get(pk=self.id)
+            Evaluation.objects.select_for_update().filter(submission_run_id=self.id)
+            if remote and submission_run.status == Status.PENDING_REMOTE:
+                SubmissionRun.objects.filter(pk=self.id).update(status=Status.CANCELLED)
```

Comment:
why aren't you doing:
```
self.status=Status.CANCELLED
self.save(update_fields=["status"]
```

Reply:
I think you can define the querysets up top, e.g.
`srun_qset = SubmissionRun.objects.filter(pk=self.id)`
and then reuse it